extensions/wikia/Captcha/Captcha.class.php

```diff
@@ -0,0 +1,622 @@
+			$wg->Request->setVal( 'wpCaptchaId', $wg->Request->getVal( 'captchaid' ) );
+		}
+		if ( $wg->Request->getVal( 'captchaword' ) ) {
+			$wg->Request->setVal( 'wpCaptchaWord', $wg->Request->getVal( 'captchaword' ) );
+		}
+		if ( $this->shouldCheck( $editPage, $newText, $section, $merged ) ) {
+			return Factory\Module::getInstance()->passCaptcha();
+		} else {
+			$this->log( "no need to show captcha.\n" );
+			return true;
+		}
+	}
+
+	/**
+	 * The main callback run on edit attempts.
+	 *
+	 * @param \EditPage $editPage
+	 * @param string $newText
+	 * @param string $section
+	 * @param bool $merged
+	 *
+	 * @return bool true to continue saving, false to abort and show a captcha form
+	 */
+	public function confirmEdit( $editPage, $newText, $section, $merged = false ) {
+		if ( defined( 'MW_API' ) ) {
+			# API mode
+			# The CAPTCHA was already checked and approved
+			return true;
+		}
+
+		$result = null;
+		$hookParams = [ &$this, &$editPage, $newText, $section, $merged, &$result ];
+		if ( !wfRunHooks( 'ConfirmEdit::onConfirmEdit', $hookParams ) ) {
+			return $result;
+		}
+
+		if ( !$this->doConfirmEdit( $editPage, $newText, $section, $merged ) ) {
+			$editPage->showEditForm( [ &$this, 'editCallback' ] );
+			return false;
+		}
+		return true;
+	}
+
+	/**
+	 * Insert the captcha prompt into an edit form.
+	 *
+	 * @param \OutputPage $out
+	 */
+	public function editCallback( &$out ) {
+		$out->addWikiText( $this->captcha->getMessage( $this->action ) );
+		$out->addHTML( $this->captcha->getForm() );
+	}
+
+	/**
+	 * A more efficient edit filter callback based on the text after section merging
+	 *
+	 * @param \EditPage $editPage
+	 * @param string $newText
+	 *
+	 * @return bool
+	 */
+	public function confirmEditMerged( $editPage, $newText ) {
+		return $this->confirmEdit( $editPage, $newText, false, true );
+	}
+
+	public function confirmEditAPI( $editPage, $newText, &$resultArr ) {
+		if ( !$this->doConfirmEdit( $editPage, $newText, false, false ) ) {
+			$this->captcha->addCaptchaAPI( $resultArr );
+			return false;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Hook for user creation form submissions.
+	 *
+	 * @param \User $u
+	 * @param string $message
+	 *
+	 * @return bool true to continue, false to abort user creation
+	 */
+	public function confirmUserCreate( $u, &$message ) {
+		$wg = \F::app()->wg;
+		if ( $wg->CaptchaTriggers['createaccount'] ) {
+			if ( $this->isIPWhitelisted() )
+				return true;
+
+			$this->trigger = "new account '" . $u->getName() . "'";
+			if ( !$this->captcha->passCaptcha() ) {
+				$message = wfMsg( 'captcha-createaccount-fail' );
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Hook for user login form submissions.
+	 *
+	 * @param \User $u
+	 * @param $pass
+	 * @param $retval
+	 *
+	 * @return bool true to continue, false to abort user creation
+	 */
+	public function confirmUserLogin( $u, $pass, &$retval ) {
+		if ( $this->isBadLoginTriggered() ) {
+			if ( $this->isIPWhitelisted() )
```

Comment:
If we're making sure to put braces around all our if statements, we should add one here as well.